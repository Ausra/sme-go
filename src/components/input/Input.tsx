import { FunctionComponent, useState } from "react";
import { defaultTheme } from "../../utils/global-styles";
import Text from "../text";
import {
  Container,
  ErrorMessageContainer,
  StyledInput,
  StyledLabel,
} from "./styles";

export enum InputStatus {
  error = "error",
}

export type InputType = "text" | "email" | "number" | "tel";
export interface InputProps extends React.HTMLAttributes<HTMLElement> {
  dataTestId?: string;
  label: string;
  type?: InputType;
  name?: string;
  status?: InputStatus;
  statusMessage?: string;
  customValue?: string | number;
  disabled?: boolean;
  pattern?: string;
  minLength?: number;
  maxLength?: number;
  onChangeCallback?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClickCallback?: (event: React.MouseEvent<HTMLInputElement>) => void;
  onBlurCallback?: (event: React.FocusEvent<HTMLInputElement>) => void;
}

export const defaultTestId = "styled-input-container";
export const errorMessageTestId = "styled-input-error-message";
export const inputTestId = "styled-input";

const Input: FunctionComponent<InputProps> = ({
  dataTestId,
  label,
  type = "text",
  onChangeCallback,
  onClickCallback,
  onBlurCallback,
  name,
  customValue,
  status,
  statusMessage,
  disabled,
  pattern,
  minLength,
  maxLength,
}) => {
  const [value, setValue] = useState("");
  const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setValue(event.target.value);
    onChangeCallback && onChangeCallback(event);
  };

  const handleOnClick = (event: React.MouseEvent<HTMLInputElement>) => {
    onClickCallback && onClickCallback(event);
  };

  const handleOnBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    onBlurCallback && onBlurCallback(event);
  };
  return (
    <Container data-testid={dataTestId || defaultTestId}>
      <StyledInput
        id="styled-input"
        data-testid={inputTestId}
        type={type}
        status={status}
        onChange={handleOnChange}
        onClick={handleOnClick}
        placeholder=" "
        name={name}
        value={customValue || value}
        onBlur={handleOnBlur}
        disabled={disabled}
        pattern={pattern}
        minLength={minLength}
        maxLength={maxLength}
      />
      <StyledLabel htmlFor="styled-input">{label}</StyledLabel>
      {status && (
        <ErrorMessageContainer data-testid={errorMessageTestId}>
          <Text color={defaultTheme.status.errorColor} size="small">
            {statusMessage}
          </Text>
        </ErrorMessageContainer>
      )}
    </Container>
  );
};

export default Input;
