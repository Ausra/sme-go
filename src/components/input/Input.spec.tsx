import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "jest-styled-components";
import Input, {
  defaultTestId,
  errorMessageTestId,
  InputStatus,
  inputTestId,
} from "./Input";

describe("Input styles", () => {
  it("renders", () => {
    render(<Input label="Company" />);
    const input = screen.getByTestId(defaultTestId);
    expect(input).toMatchSnapshot();
  });

  it("renders with errors message", () => {
    render(
      <Input
        label="Company"
        status={InputStatus.error}
        statusMessage={"error"}
      />
    );
    const errorMessage = screen.getByTestId(errorMessageTestId);
    expect(errorMessage).toMatchSnapshot();
  });
});

describe("Input", () => {
  it("renders", () => {
    render(<Input label="Company" />);
    const input = screen.getByTestId(defaultTestId);
    expect(input).toBeInTheDocument();
  });

  it("renders with errors message", () => {
    render(
      <Input
        label="Company"
        status={InputStatus.error}
        statusMessage={"error"}
      />
    );
    const errorMessage = screen.getByTestId(errorMessageTestId);
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveTextContent("error");
  });

  it("should call onChangeCallback", async () => {
    const user = userEvent.setup();
    const onChangeCallback = jest.fn();
    render(<Input label="Company" onChangeCallback={onChangeCallback} />);
    const input = screen.getByTestId(inputTestId);
    await user.type(input, "hello");
    await user.tab();

    await waitFor(() => {
      expect(onChangeCallback).toHaveBeenCalled();
    });
  });

  it("should call onClickCallback", async () => {
    const user = userEvent.setup();
    const onClickCallback = jest.fn();
    render(<Input label="Company" onClickCallback={onClickCallback} />);
    const input = screen.getByTestId(inputTestId);
    await user.click(input);

    await waitFor(() => {
      expect(onClickCallback).toHaveBeenCalled();
    });
  });

  it("should call onBlurCallback", async () => {
    const user = userEvent.setup();
    const onBlurCallback = jest.fn();
    render(<Input label="Company" onBlurCallback={onBlurCallback} />);
    const input = screen.getByTestId(inputTestId);
    await user.click(input);
    await user.tab();

    await waitFor(() => {
      expect(onBlurCallback).toHaveBeenCalled();
    });
  });
});
