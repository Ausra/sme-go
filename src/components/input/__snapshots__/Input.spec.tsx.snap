// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input styles renders 1`] = `
.c0 {
  position: relative;
  min-height: 60px;
  margin: 6px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  font-size: 14px;
  border: 1px solid;
  border-color: #CCCCCC;
  width: 100%;
  height: 60px;
  border-radius: 8px;
  padding-top: 16px;
  padding-left: 16px;
  background-color: #ffffff;
  outline: none;
}

.c2::-webkit-input-placeholder {
  color: transparent;
}

.c2::-moz-placeholder {
  color: transparent;
}

.c2:-ms-input-placeholder {
  color: transparent;
}

.c2::placeholder {
  color: transparent;
}

.c2:focus {
  background-color: #F7F7F7;
}

.c4 {
  position: absolute;
  pointer-events: none;
  left: 20px;
  top: 20px;
  -webkit-transition: 0.3s ease all;
  transition: 0.3s ease all;
}

.c1:not(:placeholder-shown) ~ .c3,
.c1:focus ~ .c4 {
  top: 10px;
  left: 16px;
  font-size: 12px;
  opacity: 0.6;
}

<div
  class="c0"
  data-testid="styled-input-container"
>
  <input
    class="c1 c2"
    data-testid="styled-input"
    id="styled-input"
    placeholder=" "
    type="text"
    value=""
  />
  <label
    class="c3 c4"
    for="styled-input"
  >
    Company
  </label>
</div>
`;

exports[`Input styles renders with errors message 1`] = `
.c1 {
  font-size: 12px;
  color: #FF7064;
  pointer-events: none;
}

.c0 {
  margin: -6px 0 6px 16px;
  height: 20px;
}

<div
  class="c0"
  data-testid="styled-input-error-message"
>
  <p
    class="c1"
    color="#FF7064"
    data-testid="styled-text"
  >
    error
  </p>
</div>
`;
